synapse/handlers/sync.py

```diff
@@ -1730,7 +1864,11 @@ async def _get_rooms_changed(
                 continue
 
             if room_id in sync_result_builder.joined_room_ids or has_join:
-                old_state_ids = await self.get_state_at(room_id, since_token)
+                old_state_ids = await self.get_state_at(
+                    room_id,
+                    since_token,
+                    state_filter=StateFilter.from_types([(EventTypes.Member, user_id)]),
+                )
                 old_mem_ev_id = old_state_ids.get((EventTypes.Member, user_id), None)
```

Comment:
We only ever pull this one local membership out of the state, so the state filter is okay to add. The same applies to the next instance below.